### PR TITLE
Feature/custom get data transfer items

### PR DESCRIPTION
**Pull request by MarinaZadoyanchuk**
**What kind of change does this PR introduce?**

- [ ] bugfix
- [x] feature
- [ ] refactoring / style
- [ ] build / chore
- [ ] documentation

**Did you add tests for your changes?**

- [ ] Yes, my code is well tested
- [ ] Not relevant

**If relevant, did you update the documentation?**

- [ ] Yes, I've updated the documentation
- [ ] Not relevant

**Summary**

Here is a very basic idea of how plugin interface could be implemented (as was suggested by @okonet in #609), basically we are giving the user an ability to specify it's own function getDataTransferItems  instead of default one. Would be great to hear some thoughts and suggestions on this.

**Does this PR introduce a breaking change?**
No

### CODE REVIEW

**Comment by okonet** [comment]
Great work on this! I really like how it's shaping.

- [ ] Please reformat the code so CI is green. It should have happen during the pre-commit hook. Did you skip it?
- [ ] The documentation needs to be updated to explain the change and we'll need example with reading directories.

**Comment by okonet** [comment]
Inviting @quarklemotion to review and comment. Also, it would be great to sync the release of the directories reader to fit the API.

**Comment by quarklemotion** [comment]
@okonet this implementation makes sense to me.  I like the use of `Promise.resolve(this.props.getDataTransferItems(evt))` which allows for both synchronous and async implementations of props.getDataTransferItems.  With the implementation in this pull request, I don't think there are any changes required to `html5-file-selector` to support this plugin interface - the `getDroppedOrSelectedFiles(event)` exported function can be used as the `getDataTransferItems` prop.  I'm happy to add a new exported function `getDroppedOrSelectedFilesWithPreview(event)` to the file selector module if everyone thinks that is the best place to locate this wrapper function tp provide preview support?

**Comment by okonet** [comment]
I think the preview plugin should live in this repo. I’d even say your plugin should probably too. I’m not sure how much overhead setting up multiple packages would be, though. Yarn workspaces should solve it, right?

**Comment by quarklemotion** [comment]
@okonet Ah, I see - I think that could make sense to convert react-dropzone to multi-package.  Were you thinking that the folder drop plugin package would be a small wrapper of html5-file-selector, or html5-file-selector would move into the react-dropzone repo?  One thing to consider is that the file selector module was built to support both dropzones and file input fields, whereas react-dropzone only cares about the dropzone support.  There are other projects using the file selector module, and I want to make sure they can continue to use it.  As for multi-package setup, I haven't used yarn workspaces before, only lerna, but happy to help with this effort!

**Comment by okonet** [comment]
Totally makes sense to me. Let’s keep thing simple for now and add your module to documentation.

File: src/index.js [hunk 1]
```diff
@@ -585,5 +606,6 @@ Dropzone.defaultProps = {
   disableClick: false,
   multiple: true,
   maxSize: Infinity,
-  minSize: 0
+  minSize: 0,
+  getDataTransferItems: defaultGetDataTransferItem
```
[the remark this answers is not in the document]
Comment (MarinaZadoyanchuk) [reply src/index.js]: I have described it in the propTypes section, should I describe it here too?

File: src/index.js [hunk 2]
```diff
@@ -156,54 +162,57 @@ class Dropzone extends React.Component {
...
-      rejectedFiles.push(...acceptedFiles.splice(1))
-    }
+    Promise.resolve(getDataTransferItems(evt)).then(fileList => {
+      const acceptedFiles = []
+      const rejectedFiles = []
+
+      fileList.forEach(file => {
+        if (!disablePreview) {
+          try {
+            file.preview = window.URL.createObjectURL(file) // eslint-disable-line no-param-reassign
```
[the remark this answers is not in the document]
Comment (Nodman) [reply src/index.js]: t’s not clear to me about how `getDataTransferItemsWithPreview`  should be separated into an additional plugin. In the previous implementation, we use to have `disablePreview` prop which would enable/disable preview generation, so if we will take this code away, how would the user be able to generate previews with a plugin? I mean, importing plugin and passing it as `getDataTransferItems` will replace original function with all the core functionality, in this case, we should also export default `getDataTransferItems` plugin so the user can make a composition of those, or maybe add an additional prop that will take function to invoke with file list after `getDataTransferItems` ?

Sorry if I am missing something!

[the remark this answers is not in the document]
Comment (okonet) [reply src/index.js]: @Nodman I think exporting the default implementation + additional plugins is a way to go. Composition should be the right way of doing this. We should also add examples of how to achieve this.

File: src/index.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (okonet) [reply src/index.js]: Oh sorry I thought that is the `propTypes`. No, the way you did it is perfect! Thanks!

**Comment by okonet** [comment]
I think the most important part we're missing is the documentation. Can you please add it and we could merge it as is. After that we could extract bits for preview.

**Comment by MarinaZadoyanchuk** [comment]
Ok, I'm going to write a usage of `getDataTransferItems` soon.

**Comment by okonet** [comment]
@MarinaZadoyanchuk thanks!

**Comment by dejongch** [comment]
Any update on when this might be merged in?

**Comment by arunchouhan163** [comment]
Please merge this PR. ASAP.

Thanks.

**Comment by okonet** [comment]
Since this one is stalling, I decided to proceed and merge it. Anyone wants to work on the documentation? @quarklemotion do you want to add an example?

**Comment by ralexrdz** [comment]
I'm not succeding implementing `getDataTransferItems`. Do you have a example I can follow?

**Comment by rolandjitsu** [comment]
@ralexrdz here you go: [using-third-party-plugins](https://react-dropzone.js.org/#using-third-party-plugins).

**Comment by ralexrdz** [comment]
@rolandjitsu thanks a lot. It was the Promise / async I was missing ;)

**Pull request merged**